### PR TITLE
Add documentation examples for `exec-used`

### PR DESCRIPTION
**Pull request by julthep**
Co-authored-by: Vladyslav Krylasov <vladyslav.krylasov@gmail.com>

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :scroll: Docs          |

## Description
Add documentation examples for `exec-used`

Refs #5953

### CODE REVIEW

File: doc/data/messages/e/exec-used/bad.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+program = input('Enter code to be executed: ')
+exec(program)  # [exec-used]
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/e/exec-used/bad.py:2]:
```suggestion
code_to_execute = """print("Hello, World!")""""
exec(code_to_execute)
```

```suggestion
code_to_execute = """input('Enter code to be executed: ')""""
program = exec(code_to_execute)  # [exec-used]
exec(program)  # [exec-used]
```

File: doc/data/messages/e/exec-used/good.py [hunk 2]
```diff
@@ -1 +1,4 @@
-# This is a placeholder for correct code for this message.
+programs = {'do_something': lambda: print("Do something")}
+program = input('Enter a program code to be used: ')
+if programs.get(program):
+    programs[program]()
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/e/exec-used/good.py:4]: I don't understand the good example. Shouldn't user inputs be one of the reason to use exec ?

```suggestion
program = input('Enter code to be executed: ')
exec(program)  # pylint: disable=exec-used
```

File: pylint/checkers/base/basic_checker.py [hunk 3]
```diff
@@ -168,7 +168,8 @@ class BasicChecker(_BasicChecker):
             "exec-used",
             'Used when you use the "exec" statement (function for Python '
             "3), to discourage its usage. That doesn't "
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/basic_checker.py:170]:
```suggestion
            'Raised when the "exec" statement is used. It's dangerous to use this function'
            " for a user input and it's also slower than actual code in general. This doesn't "
            " mean you should never use it, but you should consider alternatives first."
```

File: pylint/checkers/base/basic_checker.py [hunk 4]
```diff
@@ -168,7 +168,8 @@ class BasicChecker(_BasicChecker):
...
-            "mean you cannot use it !",
+            "mean you cannot use it! "
+            "It's dangerous to use this function for a user input.",
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/basic_checker.py:172]:
```suggestion
```

File: doc/data/messages/e/exec-used/good.py (hunk 2 above)
Comment (julthep) [reply doc/data/messages/e/exec-used/good.py:4]: I'm not so sure about them, these good/bad examples come from `pylint-errors <https://vald-phoenix.github.io/pylint-errors/plerr/errors/basic/W0122>`_.

File: pylint/checkers/base/basic_checker.py (hunk 4 above)
Comment (julthep) [reply pylint/checkers/base/basic_checker.py]: Yes, looks better. The `basic_checker.py` has been updated with your text.

File: pylint/checkers/base/basic_checker.py (hunk 3 above)
Comment (julthep) [reply pylint/checkers/base/basic_checker.py]: Yes, looks better. The `basic_checker.py` has been updated with your text.

File: doc/data/messages/e/exec-used/good.py (hunk 2 above)
Comment (Pierre-Sassoulas) [reply doc/data/messages/e/exec-used/good.py:4]: Ok, let's ditch pylint-errors example then. In the suggestion I made I wanted to show that calling code directly is better, but maybe it's a bit naive, if we have to create a function it's clearer that dynamic code creation is not ideal performance wise. Let me update:

```suggestion
def get_user_code(username):
    return input(f'Enter code to be executed please, {username}: ')

username = "Ada"
exec(get_user_code(username ))  # pylint: disable=exec-used
```

File: doc/data/messages/e/exec-used/bad.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply doc/data/messages/e/exec-used/bad.py:2]:
```suggestion
username = "Ada"
code_to_execute = f"""input('Enter code to be executed please, {username}: ')""""
program = exec(code_to_execute)  # [exec-used]
exec(program)  # [exec-used]
```

File: pylint/checkers/base/basic_checker.py [hunk 5]
```diff
@@ -166,9 +166,10 @@ class BasicChecker(_BasicChecker):
         "W0122": (
             "Use of exec",
             "exec-used",
-            'Used when you use the "exec" statement (function for Python '
-            "3), to discourage its usage. That doesn't "
-            "mean you cannot use it !",
+            'Raised when the "exec" statement is used. '
+            "It's dangerous to use this function for a user input, "
+            "and it's also slower than actual code in general. This doesn't "
+            "mean you should never use it, but you should consider alternatives first.",
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/basic_checker.py:172]:
```suggestion
            "Raised when the 'exec' statement is used. It's dangerous to use this "
            "function for a user input, and it's also slower than actual code in "
            "general. This doesn't mean you should never use it, but you should "
            "consider alternatives first.",
```

File: doc/data/messages/e/exec-used/details.rst [hunk 6]
```diff
@@ -1 +0,0 @@
-You can help us make the doc better `by contributing <https://github.com/PyCQA/pylint/issues/5953>`_ !
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/e/exec-used/details.rst:1]: The available methods and variables used in ``exec()`` may introduce a security hole. You can restrict the use of these variables and methods by passing optional globals and locals parameters (dictionaries) to the  ``exec()`` method.

File: doc/data/messages/e/exec-used/good.py [hunk 7]
```diff
@@ -1 +1,5 @@
-# This is a placeholder for correct code for this message.
+def get_user_code(username):
+    return input(f'Enter code to be executed please, {username}: ')
+
+username = "Ada"
+exec(get_user_code(username))  # pylint: disable=exec-used
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/e/exec-used/good.py:5]:
```suggestion
globals = {'__builtins__' : None}
locals = {'print': print}
exec(get_user_code(username), globals, locals)  # pylint: disable=exec-used
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: doc/data/messages/e/exec-used/details.rst (hunk 6 above)
Comment (julthep) [reply doc/data/messages/e/exec-used/details.rst:1]: `details.rst` has been created with your text.

File: doc/data/messages/e/exec-used/good.py [hunk 8]
```diff
@@ -1 +1,7 @@
-# This is a placeholder for correct code for this message.
+def get_user_code(username):
+    return input(f'Enter code to be executed please, {username}: ')
+
+username = "Ada"
+globals = {'__builtins__' : None}
+locals = {'print': print}
+exec(get_user_code(username), globals, locals)  # pylint: disable=exec-used
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/e/exec-used/good.py:7]:
```suggestion
def get_user_code(name):
    return input(f'Enter code to be executed please, {name}: ')


username = "Ada"
allowed_globals = {'__builtins__' : None}
allowed_locals = {'print': print}
exec(get_user_code(username), allowed_globals, allowed_locals)  # pylint: disable=exec-used
```

**Review by Pierre-Sassoulas — APPROVED** [review]
👍 I opened #7039 as a follow-up.

File: doc/data/messages/e/exec-used/bad.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+username = "Ada"
+code_to_execute = f"""input('Enter code to be executed please, {username}: ')""""
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/e/exec-used/bad.py:2]:
```suggestion
code_to_execute = f"""input('Enter code to be executed please, {username}: ')"""
```

File: pylint/checkers/base/basic_checker.py [hunk 10]
```diff
@@ -166,9 +166,10 @@ class BasicChecker(_BasicChecker):
         "W0122": (
             "Use of exec",
             "exec-used",
-            'Used when you use the "exec" statement (function for Python '
-            "3), to discourage its usage. That doesn't "
-            "mean you cannot use it !",
+            "Raised when the 'exec' statement is used. It's dangerous to use this "
+            "function for a user input, and it's also slower than actual code in "
+            "general. This doesn't mean you should never use it, but you should "
+            "consider alternatives first.",
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/basic_checker.py:172]:
```suggestion
            "general. This doesn't mean you should never use it, but you should "
            "consider alternatives first and restrict the functions available.",
```

**Pull request merged**